### PR TITLE
Type ascriptions

### PR DESCRIPTION
**Pull request by cquiroz**
This PR is towards scala 3 support. Scala 3 require `implicit` to have an explicit type ascription and this is included in this PR

### CODE REVIEW

File: core/src/main/scala/spire/math/Number.scala [hunk 1]
```diff
@@ -609,7 +609,14 @@ private[math] case class RationalNumber(n: Rational) extends Number { lhs =>
 }
 
 trait NumberInstances {
-  implicit final val NumberAlgebra = new NumberAlgebra
+  implicit final val NumberAlgebra: Field.WithDefaultGCD[Number]
```
Comment (milessabin) [inline core/src/main/scala/spire/math/Number.scala:612]: Why isn't this `implicit final val NumberAlgebra: NumberAlgebra = new NumberAlgebra`?

Comment (cquiroz) [reply core/src/main/scala/spire/math/Number.scala:612]: I'm following the style of cats like:
https://github.com/typelevel/cats/blob/main/kernel/src/main/scala/cats/kernel/instances/BigIntInstances.scala#L5

Comment (milessabin) [reply core/src/main/scala/spire/math/Number.scala:612]: Fair enough ... I don't know why it's done that way there.

**Pull request merged**